From 203a7b239e54f326e1ad5ff6a5d83df31c9afe2a Mon Sep 17 00:00:00 2001
From: Yi-Chia Hsieh <Yi-Chia.Hsieh@mediatek.com>
Date: Thu, 24 Jul 2025 23:02:04 +0000
Subject: [PATCH 213/254] mtk: hostapd: add TWT IE for Broadcast TWT

Add support for broadcast TWT
1. Add TWT IE in Beacon and Probe resp
2. Pass down the TWT IE offset in Beacon for offload

Signed-off-by: Yi-Chia Hsieh <Yi-Chia.Hsieh@mediatek.com>
---
 src/ap/beacon.c              | 10 ++++++
 src/ap/ieee802_11.c          |  1 +
 src/ap/ieee802_11.h          |  2 ++
 src/ap/ieee802_11_he.c       | 67 ++++++++++++++++++++++++++++++++++++
 src/common/ieee802_11_defs.h | 36 +++++++++++++++++++
 5 files changed, 116 insertions(+)

diff --git a/src/ap/beacon.c b/src/ap/beacon.c
index cfb2ea45a..c06b79abd 100644
--- a/src/ap/beacon.c
+++ b/src/ap/beacon.c
@@ -933,6 +933,9 @@ static u8 * hostapd_probe_resp_fill_elems(struct hostapd_data *hapd,
 		pos = hostapd_eid_he_capab(hapd, pos, IEEE80211_MODE_AP);
 		pos = hostapd_eid_he_operation(hapd, pos);
 
+		/* Broadcast Target Wake Time */
+		pos = hostapd_eid_twt(hapd, pos);
+
 		/* BSS Color Change Announcement element */
 		cca_pos = hostapd_eid_cca(hapd, pos);
 		if (cca_pos != pos)
@@ -2309,6 +2312,10 @@ int ieee802_11_build_ap_params(struct hostapd_data *hapd,
 
 	tail_len += he_elem_len(hapd);
 
+#ifdef CONFIG_IEEE80211AX
+	if (hapd->iconf->ieee80211ax && !hapd->conf->disable_11ax)
+		tail_len += hostapd_eid_twt_len(hapd);
+#endif /* CONFIG_IEEE80211AX */
 #ifdef CONFIG_IEEE80211BE
 	if (hapd->iconf->ieee80211be && !hapd->conf->disable_11be) {
 		tail_len += hostapd_eid_eht_capab_len(hapd, IEEE80211_MODE_AP);
@@ -2506,6 +2513,9 @@ int ieee802_11_build_ap_params(struct hostapd_data *hapd,
 			hapd->cca_c_off_beacon = cca_pos - tail - 2;
 		tailpos = cca_pos;
 
+		/* Broadcast Target Wake Time */
+		tailpos = hostapd_eid_twt(hapd, tailpos);
+
 		tailpos = hostapd_eid_spatial_reuse(hapd, tailpos);
 		tailpos = hostapd_eid_he_mu_edca_parameter_set(hapd, tailpos);
 		tailpos = hostapd_eid_he_6ghz_band_cap(hapd, tailpos);
diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index d85ef0184..6a3135805 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -5032,6 +5032,7 @@ void ieee80211_ml_build_assoc_resp(struct hostapd_data *hapd,
 	if (hapd->iconf->ieee80211ax && !hapd->conf->disable_11ax) {
 		p = hostapd_eid_he_capab(hapd, p, IEEE80211_MODE_AP);
 		p = hostapd_eid_he_operation(hapd, p);
+		p = hostapd_eid_twt(hapd, p);
 		p = hostapd_eid_spatial_reuse(hapd, p);
 		p = hostapd_eid_he_mu_edca_parameter_set(hapd, p);
 		p = hostapd_eid_he_6ghz_band_cap(hapd, p);
diff --git a/src/ap/ieee802_11.h b/src/ap/ieee802_11.h
index 4067e4a42..c82cfc7c8 100644
--- a/src/ap/ieee802_11.h
+++ b/src/ap/ieee802_11.h
@@ -173,6 +173,8 @@ bool hostapd_get_ht_vht_twt_responder(struct hostapd_data *hapd);
 void hostapd_wfa_capab(struct hostapd_data *hapd, struct sta_info *sta,
 		       const u8 *pos, const u8 *end);
 u8 * hostapd_eid_cca(struct hostapd_data *hapd, u8 *eid);
+size_t hostapd_eid_twt_len(struct hostapd_data *hapd);
+u8 * hostapd_eid_twt(struct hostapd_data *hapd, u8 *eid);
 void hostapd_tx_status(struct hostapd_data *hapd, const u8 *addr,
 		       const u8 *buf, size_t len, int ack);
 void ieee802_11_rx_from_unknown(struct hostapd_data *hapd, const u8 *src,
diff --git a/src/ap/ieee802_11_he.c b/src/ap/ieee802_11_he.c
index 0e16ea77b..7a945257d 100644
--- a/src/ap/ieee802_11_he.c
+++ b/src/ap/ieee802_11_he.c
@@ -567,3 +567,70 @@ u8 * hostapd_eid_cca(struct hostapd_data *hapd, u8 *eid)
 
 	return eid;
 }
+
+size_t hostapd_eid_twt_len(struct hostapd_data *hapd)
+{
+	size_t len;
+
+	if (hapd->broadcast_twt_cnt == 0)
+		return 0;
+
+	/* Element ID: 1 octet
+	 * Length: 1 octet
+	 * Control field: 1 octet
+	 * TWT Parameter Information: 9 octets per BTWT service period
+	 * 	Request Type: 2 octets
+	 * 	Target Wake Time: 2 octets
+	 * 	Nominal Minimum TWT Wake Duration: 1 octet
+	 * 	TWT Wake Inverval Mantissa: 2 octets
+	 * 	Broadcast TWT info subfield: 2 octets
+	 */
+	len = 2 + 1 + 9 * hapd->broadcast_twt_cnt;
+
+	return len;
+}
+
+u8 * hostapd_eid_twt(struct hostapd_data *hapd, u8 *eid)
+{
+	struct broadcast_twt_param *param;
+	u8 *pos = eid;
+	u8 *control;
+	u8 cnt = 0;
+
+	if (hapd->broadcast_twt_cnt == 0)
+		return eid;
+
+	*pos++ = WLAN_EID_TWT;
+	*pos++ = 1 + hapd->broadcast_twt_cnt * sizeof(struct btwt_param_set_field); /* len */
+	control = pos++;
+
+	*control = TWT_CONTROL_NEG_TYPE_BROADCAST | TWT_CONTROL_RX_DISABLED;
+
+	dl_list_for_each(param, &hapd->btwt_sp_list,
+			 struct broadcast_twt_param, list) {
+		struct btwt_param_set_field *btwt_ie = (struct btwt_param_set_field *) pos;
+		u16 req_type = 0, btwt_info_val = 0;
+
+		req_type = TWT_SETUP_CMD_ACCEPT << 1;
+		req_type |= param->trigger << 4;
+		req_type |= param->flow_type << 6;
+		req_type |= param->recommendation << 7;
+		req_type |= param->wake_intv_exp << 10;
+
+		cnt++;
+		if (cnt == hapd->broadcast_twt_cnt)
+			req_type |= TWT_REQTYPE_LAST_BCAST_PARAM;
+
+		btwt_ie->req_type = host_to_le16(req_type);
+		btwt_ie->min_wake_dur = param->min_wake_dur;
+		btwt_ie->wake_intv_mantissa = host_to_le16(param->wake_intv_mantissa);
+
+		btwt_info_val = (param->id << 3);
+		btwt_info_val |= (param->persist << 8);
+		btwt_ie->btwt_info = host_to_le16(btwt_info_val);
+
+		pos += sizeof(*btwt_ie);
+	}
+
+	return pos;
+}
diff --git a/src/common/ieee802_11_defs.h b/src/common/ieee802_11_defs.h
index 1da8dad9c..fd2efca00 100644
--- a/src/common/ieee802_11_defs.h
+++ b/src/common/ieee802_11_defs.h
@@ -3345,4 +3345,40 @@ struct neighbor_report_element {
 	u8 variable[0]; /* Optional Subelements */
 } STRUCT_PACKED;
 
+enum twt_setup_cmd {
+	TWT_SETUP_CMD_REQUEST,
+	TWT_SETUP_CMD_SUGGEST,
+	TWT_SETUP_CMD_DEMAND,
+	TWT_SETUP_CMD_GROUPING,
+	TWT_SETUP_CMD_ACCEPT,
+	TWT_SETUP_CMD_ALTERNATE,
+	TWT_SETUP_CMD_DICTATE,
+	TWT_SETUP_CMD_REJECT,
+};
+
+#define TWT_CONTROL_NDP			BIT(0)
+#define TWT_CONTROL_RESPONDER_PM_MODE	BIT(1)
+#define TWT_CONTROL_NEG_TYPE_BROADCAST	BIT(3)
+#define TWT_CONTROL_RX_DISABLED		BIT(4)
+#define TWT_CONTROL_WAKE_DUR_UNIT	BIT(5)
+
+#define TWT_REQTYPE_REQUEST		BIT(0)
+#define TWT_REQTYPE_SETUP_CMD		(BIT(1) | BIT(2) | BIT(3))
+#define TWT_REQTYPE_TRIGGER		BIT(4)
+#define TWT_REQTYPE_IMPLICIT		BIT(5)
+#define TWT_REQTYPE_LAST_BCAST_PARAM	BIT(5)
+#define TWT_REQTYPE_FLOWTYPE		BIT(6)
+#define TWT_REQTYPE_FLOWID		(BIT(7) | BIT(8) | BIT(9))
+#define TWT_REQTYPE_RECOMMENDATION	(BIT(7) | BIT(8) | BIT(9))
+#define TWT_REQTYPE_WAKE_INT_EXP	(BIT(10) | BIT(11) | BIT(12) | BIT(13) | BIT(14))
+#define TWT_REQTYPE_PROTECTION		BIT(15)
+
+struct btwt_param_set_field {
+	le16 req_type;
+	le16 twt;
+	u8 min_wake_dur;
+	le16 wake_intv_mantissa;
+	le16 btwt_info;
+} STRUCT_PACKED;
+
 #endif /* IEEE802_11_DEFS_H */
-- 
2.45.2

